sno/working_copy/base.py

```diff
@@ -452,15 +642,157 @@ def update_state_table_tree(self, tree):
+                )
+                L.info(
+                    "Overall rate: %d features/s", (feat_progress / (t1 - t0 or 0.001))
+                )
+
+                self._create_triggers(db, dataset)
+                self._update_last_write_time(db, dataset, commit)
+
+            db.execute(
+                f"""
+                INSERT INTO {self.SNO_STATE} (table_name, key, value) VALUES ('*', 'tree', :value)
+                ON CONFLICT (table_name, key) DO UPDATE SET value=EXCLUDED.value;
+                """,
+                {"value": commit.peel(pygit2.Tree).hex},
+            )
+
+    def _create_table_for_dataset(self, db, dataset):
+        """Create the working-copy table for checking out the given dataset."""
+        raise NotImplementedError
+
+    def _write_meta(self, db, dataset):
+        """Write any non-feature data relating to dataset - title, description, CRS, etc."""
+        raise NotImplementedError()
+
+    def _create_spatial_index(self, db, dataset):
+        """
+        Creates a spatial index for the table for the given dataset.
+        The spatial index is configured so that it is automatically updated when the table is modified.
+        It is not guaranteed that the spatial index will take into account features that are already present
+        in the table when this function is called - therefore, this should be called while the table is still empty.
+        """
         raise NotImplementedError()
 
+    def _drop_spatial_index(self, db, dataset):
+        """Inverse of _create_spatial_index - deletes the spatial index."""
+        raise NotImplementedError()
+
+    def _update_last_write_time(self, db, dataset, commit=None):
+        """Hook for updating the last-modified timestamp stored for a particular dataset, if there is one."""
+        pass
+
+    def _write_features(self, db, dataset, pk_list, *, ignore_missing=False):
+        """Write the features from the dataset with the given PKs to the table for the dataset."""
+        if not pk_list:
+            return 0
+
+        sql = self.insert_or_replace_into_dataset(dataset)
+        feat_count = 0
+        CHUNK_SIZE = 10000
+        for row_dicts in self._chunk(
+            dataset.get_features_with_crs_ids(pk_list, ignore_missing=ignore_missing),
+            CHUNK_SIZE,
+        ):
+            r = db.execute(sql, row_dicts)
+            feat_count += r.rowcount
+
+        return feat_count
+
+    def _delete_features(self, db, dataset, pk_list):
+        if not pk_list:
+            return 0
+
+        pk_column = self.preparer.quote(dataset.primary_key)
+        sql = f"""DELETE FROM {self.table_identifier(dataset)} WHERE {pk_column} IN :pks;"""
+        feat_count = 0
+        CHUNK_SIZE = 100
+        for pks in self._chunk(pk_list, CHUNK_SIZE):
+            t = sqlalchemy.text(sql)
+            t = t.bindparams(sqlalchemy.bindparam("pks", expanding=True))
```

Comment:
I think you can move these lines outside the loop, saves a bit of computation

Reply:
Done